### ISSUE TITLE
Preformatted blocks are losing whitespace.

### ISSUE THREAD

**Issue by diegoreymendez** [WordPress/gutenberg#6917]
**Describe the bug**

Preformatted blocks are not really working as expected.  They're losing whitespace when converting from visual to HTML and back.

**To Reproduce**

1. Add a "Preformatted" block
2. Write some text with extensive usage of whitespace.  For example

```
This is some text
     in two lines        yes two lines
```
3. Save the post.
4. Exit and reopen it.

**Expected behavior**

The whitespace should be maintained.

**Screenshots**

<img width="667" alt="screen shot 2018-05-23 at 12 24 51" src="https://user-images.githubusercontent.com/1836005/40434390-a2706e16-5e84-11e8-9bfc-1a071bd7da4d.png">

<img width="669" alt="screen shot 2018-05-23 at 12 25 24" src="https://user-images.githubusercontent.com/1836005/40434414-ac62707c-5e84-11e8-8b9a-870ecb912f45.png">

**Desktop (please complete the following information):**
 - OS: OS X
 - Browser Safari
 - Version 2.9.2

**Additional context**
- Please add the version of Gutenberg you are using in the description.
- To report a security issue, please visit the WordPress HackerOne program: https://hackerone.com/wordpress.

**Comment by danielbachhuber** [comment]
Hi @diegoreymendez,

Thanks for the detailed report! I can't reproduce this issue in Safari Version 11.1 (13605.1.33.1.4) on master or in Gutenberg v2.9.2:

![pretext](https://user-images.githubusercontent.com/36432/40543410-10f0f4aa-5fd9-11e8-8289-e56eabde29c1.gif)

Are you running any plugins that might impact what's being saved? Can you debug further to see if you can track down any more details?

**Comment by diegoreymendez** [comment]
In the following example, the plugins I have are:

Gutenberg 2.9.2
Jetpack 6.1.1

![testinggutenberg](https://user-images.githubusercontent.com/1836005/40622065-0e8d7ad6-6276-11e8-9454-da502dc4ba3e.gif)

**Comment by danielbachhuber** [comment]
Can you try disabling Jetpack and seeing if he problem reproduces with Jetpack disabled?

**Comment by diegoreymendez** [comment]
@danielbachhuber - I'm getting the same issue with Jetpack disabled.  It's not properly rendering whitespace when following the steps from my GIF above.

**Comment by danielbachhuber** [comment]
@diegoreymendez Odd. Could it be a browser extension?

**Comment by diegoreymendez** [comment]
@danielbachhuber - This is what I see without WP Plugins (other than Gutenberg), and without Safari extensions (the only one I had was 1Password, and I disabled it for this).

![gutenproblems](https://user-images.githubusercontent.com/1836005/40921377-abc970b0-67e5-11e8-8748-8d6e89b3a68c.gif)

I noticed you tested this doing some form of page reload... can you try switching to HTML mode and back to see if you can reproduce the same behavior?

**Comment by danielbachhuber** [comment]
> can you try switching to HTML mode and back to see if you can reproduce the same behavior?

Yes, I can reproduce your bug report now. Thanks for diligently tracking down reproduction steps.

**Comment by rosswintle** [comment]
I've found what I think is the same thing. It occurs when duplicating blocks and when saving a post and re-loading.

This is:
* Chrome 67
* MacOS High Sierra 10.13.5
* WP 4.9.6
* Gutenberg 3.0.1
* Lots of plugins (this is on my live blog site)

Video: https://cl.ly/2C0d2w2W291P

Looks like you've managed to replicate but do you want me to try and make this happen in a simpler setup?

**Comment by aduth** [comment]
I expect this is likely related to / duplicate of #4456 , caused by block serializer's beautification removing excess whitespace.

Though I might have thought `js-beautify` would at least ignore text within `pre` tags. Seems it's being respected okay at least for the original block?

>I wasn't able to reproduce the aforementioned steps with the paragraph block, so it seems specific to the pre-formatted block.

It could be that TinyMCE is inserting whitespace as `&nbsp;` entities, whereas the Preformatted block is using a simple textarea field.

**Comment by danielbachhuber** [comment]
Some good examples in #7836
